File: app/src/main/java/iuliiaponomareva/eventum/adapters/NewsArrayAdapter.java
package iuliiaponomareva.eventum.adapters;

import android.app.Activity;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.support.annotation.NonNull;
import android.text.Html;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import java.util.List;

import iuliiaponomareva.eventum.R;
import iuliiaponomareva.eventum.data.News;

public class NewsArrayAdapter extends ArrayAdapter<News> {
    private Context context;

    Picasso getPicasso() {
        return picasso;
    }

    private Picasso picasso;

    String getChannelURL() {
        return channelURL;
    }

    private String channelURL;

    public NewsArrayAdapter(Context context, List<News> news) {
        super(context, R.layout.news_list_item, news);
        this.context = context;
        picasso = Picasso.with(context);
    }


    private static class ViewHolder {

        private TextView textView;
    }

    public void setChannelURL(String channelURL) {
        this.channelURL = channelURL;
    }

    public void cancel() {
        picasso.cancelTag(channelURL);
    }

    @NonNull
    @Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        ViewHolder holder;
        if (convertView == null) {
            LayoutInflater inflater = (LayoutInflater) context
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            convertView = inflater.inflate(R.layout.news_list_item, parent, false);
            holder = new ViewHolder();
            holder.textView = (TextView) convertView.findViewById(R.id.news_textview);
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            holder.textView.setText(Html.fromHtml(getItem(position).toString(),
                    Html.FROM_HTML_MODE_LEGACY,
                    new MyImageGetter(this, holder.textView), null));
        } else {
            holder.textView.setText(Html.fromHtml(getItem(position).toString(),
                    new MyImageGetter(this, holder.textView), null));
        }
        return convertView;
    }

}
class MyDrawable extends BitmapDrawable {
    MyDrawable(Resources res) {
        super(res, (Bitmap) null);
    }

    void setDrawable(Drawable drawable) {
        this.drawable = drawable;
    }

    private Drawable drawable;

    @Override
    public void draw(Canvas canvas) {
        if (drawable != null) {
            drawable.draw(canvas);
        }
    }
}

class MyImageGetter implements Html.ImageGetter, Target {
    private NewsArrayAdapter newsArrayAdapter;
    private int width;
    private String source;
    private MyDrawable image;
    private TextView view;

    MyImageGetter(NewsArrayAdapter newsArrayAdapter, TextView view) {
        super();
        this.newsArrayAdapter = newsArrayAdapter;
        this.view = view;
        DisplayMetrics metrics = new DisplayMetrics();
        ((Activity) newsArrayAdapter.getContext()).getWindowManager().getDefaultDisplay()
                .getMetrics(metrics);
        width = metrics.widthPixels;
    }

    @Override
    public Drawable getDrawable(String source) {
        image = new MyDrawable(newsArrayAdapter.getContext().getResources());
        if (source.startsWith("//")) {
            source = "http:" + source;
        }
        this.source = source;
        newsArrayAdapter.getPicasso().load(source).tag(newsArrayAdapter.getChannelURL())
                .resize(width, 0)
                .onlyScaleDown().into(this);
        return image;
    }


    @Override
    public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
        if ((source != null) && (bitmap != null)) {
            Drawable image = new BitmapDrawable(newsArrayAdapter.getContext().getResources(),
                    bitmap);
            image.setBounds(0, 0, image.getIntrinsicWidth(), image.getIntrinsicHeight());
            this.image.setBounds(0, 0, image.getIntrinsicWidth(), image.getIntrinsicHeight());
            this.image.setDrawable(image);
            view.invalidate();
            view.setText(view.getText());
        }
    }

    @Override
    public void onBitmapFailed(Drawable errorDrawable) {

    }

    @Override
    public void onPrepareLoad(Drawable placeHolderDrawable) {

    }
}

File: app/src/main/java/iuliiaponomareva/eventum/adapters/NewsArrayAdapter.kt
package iuliiaponomareva.eventum.adapters

import android.app.Activity
import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.Build
import android.text.Html
import android.text.Html.ImageGetter
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import com.squareup.picasso.Picasso
import com.squareup.picasso.Picasso.LoadedFrom
import com.squareup.picasso.Target
import iuliiaponomareva.eventum.R
import iuliiaponomareva.eventum.adapters.NewsArrayAdapter.Companion.DEFAULT_TAG
import iuliiaponomareva.eventum.data.News

class NewsArrayAdapter(
    aContext: Context,
    news: List<News>
) : ArrayAdapter<News>(aContext, R.layout.news_list_item, news) {

    val picasso: Picasso = Picasso.get()

    var channelURL: String? = null

    private class ViewHolder {
        lateinit var textView: TextView
    }

    fun cancel() {
        channelURL?.let { picasso.cancelTag(it) }
        picasso.cancelTag(DEFAULT_TAG)
    }

    override fun getView(
        position: Int,
        convertView: View?,
        parent: ViewGroup
    ): View {
        val view: View
        val holder: ViewHolder
        if (convertView == null) {
            val inflater = context
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            view = inflater.inflate(R.layout.news_list_item, parent, false)
            holder = ViewHolder()
            holder.textView = view.findViewById(R.id.news_textview)
            view.tag = holder
        } else {
            view = convertView
            holder = view.tag as ViewHolder
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            holder.textView.text = Html.fromHtml(
                getItem(position).toString(),
                Html.FROM_HTML_MODE_LEGACY,
                MyImageGetter(this, holder.textView), null
            )
        } else {
            holder.textView.text = Html.fromHtml(
                getItem(position).toString(),
                MyImageGetter(this, holder.textView), null
            )
        }
        return view
    }

    companion object {
        const val DEFAULT_TAG = "all"
    }

}

internal class MyDrawable(res: Resources) :
    BitmapDrawable(res, null as Bitmap?) {
    fun setDrawable(drawable: Drawable?) {
        this.drawable = drawable
    }

    private var drawable: Drawable? = null
    override fun draw(canvas: Canvas) {
        drawable?.draw(canvas)
    }
}

internal class MyImageGetter(
    private val newsArrayAdapter: NewsArrayAdapter,
    private val view: TextView?
) :
    ImageGetter, Target {
    private val width: Int
    private var source: String? = null
    private var image: MyDrawable? = null
    override fun getDrawable(source: String): Drawable {
        var source = source
        image = MyDrawable(newsArrayAdapter.context.resources)
        if (source.startsWith("//")) {
            source = "http:$source"
        }
        this.source = source
        newsArrayAdapter.picasso.load(source).tag(newsArrayAdapter.channelURL ?: DEFAULT_TAG)
            .resize(width, 0)
            .onlyScaleDown()
            .into(this)
        return image!!
    }

    override fun onBitmapLoaded(bitmap: Bitmap, from: LoadedFrom) {
        if (source != null && bitmap != null) {
            val image: Drawable = BitmapDrawable(
                newsArrayAdapter.context.resources,
                bitmap
            )
            image.setBounds(0, 0, image.intrinsicWidth, image.intrinsicHeight)
            this.image!!.setBounds(0, 0, image.intrinsicWidth, image.intrinsicHeight)
            this.image!!.setDrawable(image)
            view!!.invalidate()
            view.text = view.text
        }
    }

    override fun onBitmapFailed(
        exception: Exception,
        errorDrawable: Drawable
    ) {
    }

    override fun onPrepareLoad(placeHolderDrawable: Drawable?) {}

    init {
        val metrics = DisplayMetrics()
        (newsArrayAdapter.context as Activity).windowManager.defaultDisplay
            .getMetrics(metrics)
        width = metrics.widthPixels
    }
}